convert_hf_to_gguf.py

```diff
@@ -2712,6 +2713,78 @@ class StarCoder2Model(Model):
     model_arch = gguf.MODEL_ARCH.STARCODER2
 
 
+@Model.register("Rwkv6ForCausalLM")
+class RwkvModel(Model):
+    model_arch = gguf.MODEL_ARCH.RWKV6
+
+    def set_vocab(self):
+        assert (self.dir_model / "rwkv_vocab_v20230424.txt").is_file()
+        vocab_size = self.hparams.get("vocab_size", 65536)
+
+        tokens: list[bytes] = ['<s>'.encode("utf-8")]
+        toktypes: list[int] = [gguf.TokenType.CONTROL]
+
+        with open(self.dir_model / "rwkv_vocab_v20230424.txt", "r", encoding="utf-8") as f:
+            lines = f.readlines()
+            for line in lines:
+                parts = line.split(' ')
+                assert len(parts) >= 3
+                token, token_len = ast.literal_eval(' '.join(parts[1:-1])), int(parts[-1])
+                token = token.encode("utf-8") if isinstance(token, str) else token
+                assert isinstance(token, bytes)
+                assert len(token) == token_len
+                token_text: str = repr(token)[2:-1]  # "b'\xff'" -> "\xff"
+                tokens.append(token_text.encode("utf-8"))
+                toktypes.append(gguf.TokenType.NORMAL)
+        remainder = vocab_size - len(tokens)
+        assert remainder >= 0
+        for i in range(len(tokens), vocab_size):
+            tokens.append(f"[PAD{i}]".encode("utf-8"))
+            toktypes.append(gguf.TokenType.UNUSED)
+
+        self.gguf_writer.add_tokenizer_model("rwkv")
+        self.gguf_writer.add_token_list(tokens)
+        self.gguf_writer.add_token_types(toktypes)
+
+    def set_gguf_parameters(self):
+        block_count = self.hparams["num_hidden_layers"]
+        head_size = self.hparams["head_size"]
+        hidden_size = self.hparams["hidden_size"]
+        layer_norm_eps = self.hparams["layer_norm_epsilon"]
+        rescale_every_n_layers = self.hparams["rescale_every"]
+
+        # RWKV isn't context limited
+        self.gguf_writer.add_context_length(1048576)
+        self.gguf_writer.add_embedding_length(hidden_size)
+        self.gguf_writer.add_block_count(block_count)
+        self.gguf_writer.add_layer_norm_eps(layer_norm_eps)
+        self.gguf_writer.add_rescale_every_n_layers(rescale_every_n_layers)
+        self.gguf_writer.add_wkv_head_size(head_size)
+
+        # required by llama.cpp, unused
+        self.gguf_writer.add_head_count(0)
+        self.gguf_writer.add_feed_forward_length(0)
+
+    def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iterable[tuple[str, Tensor]]:
+        new_name = self.map_tensor_name(name)
+
+        if not (new_name.endswith(".weight") or new_name.endswith(".bias")):
+            new_name += ".weight"
```

Comment:
Somehow after changing ``or not name.endswith(".weight")`` with ``or not new_name.endswith(".weight")`` the converted gguf will make a CUDA error when inferencing. Gonna keep the existing one for now in this PR

Reply:
> Somehow after changing `or not name.endswith(".weight")` with `or not new_name.endswith(".weight")` the converted gguf will make a CUDA error when inferencing. Gonna keep the existing one for now in this PR

Models quantized with `llama-quantize` will *also* have this problem, since using `new_name` in the convert script simply makes it use the same types as `llama-quantize`. I think both methods should result in the same model files.

Which tensors are problematic? Maybe they need to be explicitly excluded from quantization (assuming they are small enough)?

Reply:
> > Somehow after changing `or not name.endswith(".weight")` with `or not new_name.endswith(".weight")` the converted gguf will make a CUDA error when inferencing. Gonna keep the existing one for now in this PR
> 
> Models quantized with `llama-quantize` will _also_ have this problem, since using `new_name` in the convert script simply makes it use the same types as `llama-quantize`. I think both methods should result in the same model files.
> 
> Which tensors are problematic? Maybe they need to be explicitly excluded from quantization (assuming they are small enough)?

Yes, they are not that large.
I've added the commit that matches new_name, as well as one that adds two problematic tensors in f32 list